### PR TITLE
[INFRA] Improve the java style checks log the errors to sbt console

### LINKED ISSUE #3067: [Feature Request][INFRA] Improve the java checkstyle checks to log the errors to the sbt console
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [X] Standalone
- [X] Flink
- [X] Kernel
- [ ] Other (fill in here)

### Overview

When we upgraded our `sbt` version we were forced to upgrade our the sbt java checkstyle plug in https://github.com/delta-io/delta/pull/3019. This upgrade means that the java checkstyle errors are no longer logged to the sbt console. Fixing errors requires looking at the `$projectName/target/checkstyle-report.xml` and `$projectName/target/checkstyle-test-report.xml`. This is a poor dev workflow, and much inferior to the previous state where the errors would be printed to the console.

### Further details

We can add our own reading of the checkstyle reports and logging to console as part of our sbt build. For inspiration, we can look at the code that was used in the former sbt checkstyle plugin here https://github.com/etsy/sbt-checkstyle-plugin/blob/master/src/main/scala/com/etsy/sbt/checkstyle/Checkstyle.scala#L70-L97

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [X] No. I cannot contribute this feature at this time.

### PR DESCRIPTION
**Pull request by tlm365**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [x] Standalone
- [x] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description
Resolves #3067.

## How was this patch tested?
On local machine, intentionally create checkstyle errors in module `kernelDefaults` (for experimental), then run the `build/sbt compile` and `build/sbt kernelDefaults/test`.

## Does this PR introduce _any_ user-facing changes?
No.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1335,18 +1335,51 @@ lazy val scalaStyleSettings = Seq(
...
+
+    testJavastyle := {
+      (Test / checkstyle).value
+      javaCheckstyle(streams.value.log, (Compile / target).value / "checkstyle-test-report.xml")
+    },
+    (Test / test) := ((Test / test) dependsOn (Test / testJavastyle)).value
   )
 }
 
+def javaCheckstyle(log: Logger, reportFile: File): Unit = {
```
[the remark this answers is not in the document]
Comment (tlm365) [reply build.sbt:1363]: @vkorukanti Yes, I have tried to read the two checkstyle configs `dev/kernel-checkstyle.xml` and `dev/connectors-checkstyle.xml`, and researched on configuration document [here](https://checkstyle.sourceforge.io/config.html). But I don't find any config related to log errors to the console. Seems like that feature isn't supported.

File: build.sbt [hunk 2]
```diff
@@ -1335,18 +1335,51 @@ lazy val scalaStyleSettings = Seq(
...
+      (file, line, message)
+    }
+  }
+
+  if (errors.nonEmpty) {
+    errors.foreach { case (file, line, message) =>
+      log.error(s"- File: $file, Line: $line, Message: $message")
+    }
+    sys.error(s"Found checkstyle errors")
+  }
```
Comment (vkorukanti) [inline build.sbt:1380]: It looks good to me. The only request is, can we print the errors as part of the failure message? That way we don't need to scroll up to find the error.

```
    var errorMsg = "Found checkstyle errors"
    errors.foreach { case (file, line, message) =>
      val lineError = s"File: $file, Line: $line, Message: $message"
      log.error(lineError)
      errorMsg += ("\n" + lineError)
    }
    sys.error(errorMsg + "\n")
```

**Review by vkorukanti — CHANGES_REQUESTED** [review]
LGTM. One comment. Tried locally and it works great.

File: build.sbt [hunk 3]
```diff
@@ -1335,18 +1335,51 @@ lazy val scalaStyleSettings = Seq(
...
 def javaCheckstyleSettings(checkstyleFile: String): Def.SettingsDefinition = {
   // Can be run explicitly via: build/sbt $module/checkstyle
   // Will automatically be run during compilation (e.g. build/sbt compile)
   // and during tests (e.g. build/sbt test)
   Seq(
     checkstyleConfigLocation := CheckstyleConfigLocation.File(checkstyleFile),
-    checkstyleSeverityLevel := CheckstyleSeverityLevel.Error,
-    (Compile / compile) := ((Compile / compile) dependsOn (Compile / checkstyle)).value,
-    (Test / test) := ((Test / test) dependsOn (Test / checkstyle)).value
+    checkstyleSeverityLevel := CheckstyleSeverityLevel.Ignore,
```
Comment (vkorukanti) [inline build.sbt:1347]: one naive qn: why is this changed from `Error` to `Ignore`?

Comment (tlm365) [reply build.sbt:1347]: > Is it because we want to control the part about throwing error?

@vkorukanti Yes, that's right. If we keep the `Error` severity, `build/sbt` will throw an error and immediately stop at the checkstyle phase (if error) -> never execute the check-report phase of `checkstyle-report.xml` and `checkstyle-test-report.xml`. We need to `Ignore` and throw error if exists when checking `*report.xml`.

File: build.sbt (hunk 2 above)
Comment (tlm365) [reply build.sbt:1380]: @vkorukanti Thanks for reviewing.

File: build.sbt [hunk 4]
```diff
@@ -1335,18 +1335,51 @@ lazy val scalaStyleSettings = Seq(
  ****************************
  */
 
+lazy val compileJavastyle = taskKey[Unit]("compileJavastyle")
+lazy val testJavastyle = taskKey[Unit]("testJavastyle")
+
 def javaCheckstyleSettings(checkstyleFile: String): Def.SettingsDefinition = {
```
[the remark this answers is not in the document]
Comment (tlm365) [reply build.sbt:1341]: @allisonport-db Thanks for reviewing. Your point sounds good, we have a quite complicated `build.sbt` now. But if isolating this checkstyle part to a separate file, where should we organize that file?

[the remark this answers is not in the document]
Comment (vkorukanti) [reply build.sbt:1341]: We keep some of the build-related code (e.g., Unidos) under the `project` directory.  You can put it there.

**Review by vkorukanti — APPROVED** [review]
LGTM. Thank you for this change. It makes it easy to find the `checkstyle` errors.

I made a couple of minor edits. Let me know if the changes look fine to you.

**Comment by tlm365** [comment]
> LGTM. Thank you for this change. It makes it easy to find the `checkstyle` errors.
> I made a couple of minor edits. Let me know if the changes look fine to you.

@vkorukanti  It's nice. TYSM!

**Review by allisonport-db — APPROVED** [review]
